diff --git a/adb/line_printer.cpp b/adb/line_printer.cpp
index 4ec8979..144a3ae 100644
--- a/adb/line_printer.cpp
+++ b/adb/line_printer.cpp
@@ -14,6 +14,9 @@
 
 #include "line_printer.h"
 
+// Make sure printf is really adb_printf which works for UTF-8 on Windows.
+#include <sysdeps.h>
+
 #include <stdio.h>
 #include <stdlib.h>
 #ifdef _WIN32
@@ -25,9 +28,6 @@
 #include <sys/time.h>
 #endif
 
-// Make sure printf is really adb_printf which works for UTF-8 on Windows.
-#include <sysdeps.h>
-
 // Stuff from ninja's util.h that's needed below.
 #include <vector>
 using namespace std;
