diff --git a/mDNSShared/CommonServices.h b/mDNSShared/CommonServices.h
index be49257..966a7a4 100644
--- a/mDNSShared/CommonServices.h
+++ b/mDNSShared/CommonServices.h
@@ -260,11 +260,13 @@
 		#pragma warning( disable:4127 )		// Disable "conditional expression is constant" warning for debug macros.
 		#pragma warning( disable:4706 )		// Disable "assignment within conditional expression" for Microsoft headers.
 		
+	#elif( defined (USE_MINGW) )
+		#include <stdint.h>
 	#endif
 
 	#include	<windows.h>
 	#include	<winsock2.h>
-	#include	<Ws2tcpip.h>
+	#include	<ws2tcpip.h>
 	
 	#if( defined( _MSC_VER ) )
 		#pragma warning( default:4706 )
diff --git a/mDNSWindows/DLL/dllmain.c b/mDNSWindows/DLL/dllmain.c
index 79f3bb7..0c8fb36 100644
--- a/mDNSWindows/DLL/dllmain.c
+++ b/mDNSWindows/DLL/dllmain.c
@@ -17,7 +17,7 @@
 
 #include <windows.h>
 #include <DebugServices.h>
-
+#if 0
 BOOL APIENTRY	DllMain( HANDLE inModule, DWORD inReason, LPVOID inReserved )
 {
 	(void) inModule;
@@ -33,7 +33,7 @@ BOOL APIENTRY	DllMain( HANDLE inModule, DWORD inReason, LPVOID inReserved )
 	}
     return( TRUE );
 }
-
+#endif
 
 BOOL
 IsSystemServiceDisabled()
